### ISSUE TITLE
`org.logstash.Timestamp` Seems Redundant

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#7897]
`org.logstash.Timestamp` holds no state beyond the `org.joda.time.DateTime` it wraps.
It seems the only point of having it around is to format timestamps a certain way and facilitate serialization via `org.logstash.json.TimestampSerializer`?

Both the formatting of the timestamp as well as the serialization in a certain format can be achieved by other (faster and 0GC) means from a `long`.
Is there any reason to keep this class around and not replace it with a less heavy approach?

(same goes for related logic around `org.logstash.ext.JrubyTimestampExtLibrary` and the `BiValue` we have for this one)

**Comment by jakelandis** [comment]
I don't think we can get rid of `JrubyTimestampExtLibrary` since it effectively exposes the Joda `DateTime` to Ruby. (right ?) 

It seems that `org.logstash.Timestamp` implements `org.logstash.ackedqueue.Queueable` which  Joda obviously doesn't implement. I am not sure the exact usage, but it seems like a fairly standard pattern of wrapping an object to allow it to conform to a specific interface.  It also seems to bring some consistency to the formatter and simplifies the Joda contract. 

How expensive is the `org.logstash.Timestamp` wrapper ?  

imo, it is a nice abstraction, and wouldn't want to toss it unless there are measurable performance gains.

**Comment by original-brownbear** [comment]
@jakelandis 

> I don't think we can get rid of JrubyTimestampExtLibrary since it effectively exposes the Joda DateTime to Ruby. (right ?)

Well, it does expose that to Ruby, but it's certainly not a very fast way of doing so. Since all of this stuff has no (non-redundant) state, you can just get the same effect without the memory cost with some `static` converter (by definition :P).

> It seems that org.logstash.Timestamp implements org.logstash.ackedqueue.Queueable which Joda obviously doesn't implement.

I don't think `Queueable` is even used on this thing outside of tests. Even if it is, `Queueable` is kind of a pointless interface in the first place if you think about it. This could again simply be cleaned up by a static utility with the signature `Queueable method(long var)` (should we actually need this).

> it seems like a fairly standard pattern of wrapping an object to allow it to conform to a specific interface.

Not in the hot path, especially we're just using way too much memory already, looking at how visible every small improvement is in throughput (we're still much much heavier than the old Ruby event :().

> How expensive is the org.logstash.Timestamp wrapper ?

This is impossible to answer in terms of throughput, but the profiler shows me `40` byte per timestamp, so at least `32` byte in waste per event (this isn't so relevant obviously).
The fact that I see the bulk of these instances `82%` being unreachable but not collected shows me there's something problematic going on with these (kind of hard to track it down with all the wrappers and also pointless when the faster solution of `static` converters for `long` is right in front of us).

> imo, it is a nice abstraction, and wouldn't want to toss it unless there are measurable performance gains.

Always hard to see the eventual effect of this stuff, but it's a fact that we're slower than the old Ruby Event and it has to come from this thing and other type conversions.

**Comment by jakelandis** [comment]
>bulk of these instances 82% being unreachable but not collected 
Do you mean unreachable from GC roots ? 

If so doesn't that just mean that they can be garbage collected, but just haven't been yet? Since they are tiny objects I would guess that the that GC just isn't prioritizing them.

**Comment by original-brownbear** [comment]
It's hard to quantify this, but if it can be avoided like with time-stamps and in situations where steady-state throughput matters, the cost in invalidated CPU cache from all the indirect references that play into this object and its eventual GC will likely matter.
I'll see if I can benchmark a simple case and remove the logic around the timestamp to get an estimate on this tomorrow :)

**Comment by jordansissel** [comment]
The history was this:

* Older Logstash used to monkeypatch the ruby Time class: https://github.com/elastic/logstash/blob/v1.2.0/lib/logstash/event.rb#L17
* At that time, we were publishing part of Logstash (the Event class) to be usable by other projects, and Logstash's monkeypatching of Time was problematic for Rails users who were using this Event class.
* In Logstash 1.5 we moved `@timestamp` to be a LogStash::Timestamp type which lightly wraps Ruby's Time: https://github.com/elastic/logstash/blob/v1.5.0/lib/logstash/timestamp.rb
* In Logstash 2.2.0, we moved the implementation to Java and wraps a Joda DateTime: https://github.com/elastic/logstash/blob/v2.2.0/logstash-core-event-java/src/main/java/com/logstash/Timestamp.java

**Comment by jordansissel** [comment]
@original-brownbear are you proposing the value stored in `@timestamp` field be, internally, a long (milliseconds since epoch) formatting (via event sprintf) and serialization invoking Joda each time?

**Comment by original-brownbear** [comment]
@jordansissel 

`long` would be one option, but only if it actually saves instantiations, would have to go through the logic in detail to see if that is better or worse than keeping a reference to a Joda `DateTime`.
I'm simply in favor of dropping this class because it adds absolutely nothing beyond what is already there in `DateTime`. Just more code and a potential performance hit.

**Comment by original-brownbear** [comment]
Gave it one pass here https://github.com/elastic/logstash/pull/7916, after some cleanup it seems even more redundant than before though. We already hold the Joda time object in Ruby's `RubyTime`?? So why are we even going through the whole exercise here? Seems pretty much the same as with `StringBiValue` (only that it's kind of twice as bad :D, because we also have a `TimestampBiValue` ... wrapping the wrapper)?
I guess we could just streamline all handling of the timestamp to `RubyTimestamp` + Joda's `DateTime` in exactly the same manner we handled `StringBiValue`. With `DateTime` being nicely immutable, this should have 0 allocation overhead for the Java side even unlike with `RubyString` and `String`.
